ddtrace/contrib/grpc/client_interceptor.py

```diff
@@ -0,0 +1,71 @@
+import grpc
+import collections
+
+from ddtrace import Pin
+
+class _ClientCallDetails(
+        collections.namedtuple(
+            '_ClientCallDetails',
+            ('method', 'timeout', 'metadata', 'credentials')),
+        grpc.ClientCallDetails):
+    pass
+
+class _GrpcClientInterceptor(
+        grpc.UnaryUnaryClientInterceptor, grpc.UnaryStreamClientInterceptor,
+        grpc.StreamUnaryClientInterceptor, grpc.StreamStreamClientInterceptor):
+
+    def __init__(self, host, port):
+        pin = Pin.get_from(grpc)
+        self._service = pin.service
+        self._tracer = pin.tracer
+        self._host = host
+        self._port = port
+
+    def _start_span(self, method):
+        span = self._tracer.trace('grpc.client', span_type='grpc', service=self._service, resource=method)
+        span.set_tag('grpc.host', self._host)
+        if (self._port is not None):
+            span.set_tag('grpc.port', self._port)
+        return span
+
+    def _inject_span(self, span, client_call_details):
+        metadata = []
+        if client_call_details.metadata is not None:
+            metadata = list(client_call_details.metadata)
+        metadata.append((b'x-datadog-trace-id', str(span.trace_id)))
+        metadata.append((b'x-datadog-parent-id', str(span.span_id)))
+
+        if (span.context.sampling_priority) is not None:
+            metadata.append((b'x-datadog-sampling-priority', str(span.context.sampling_priority)))
+        client_call_details = _ClientCallDetails(
+            client_call_details.method, client_call_details.timeout, metadata,
+            client_call_details.credentials)
+        return client_call_details
+
+    def intercept_unary_unary(self, continuation, client_call_details, request):
+        return self.intercept_unary_stream(continuation, client_call_details, request)
+
+    def intercept_unary_stream(self, continuation, client_call_details, request):
+        with self._start_span(client_call_details.method) as span:
```

Comment:
We should follow a pattern of:

```python
# Get the pin and ensure it exists and is enabled
pin = Pin.get_from(grpc)
if not pin or not pin.enabled():
    return continuation(client_call_details, request)

with self._start_span(...) as span:
    pass
```

Example: https://github.com/DataDog/dd-trace-py/blob/87b9ff74509fab917c7341f34ffabbbf14b74578/ddtrace/contrib/redis/patch.py#L43-L57

We want to make sure that we do not create any spans if they have it disabled.

Reply:
We might not need this pin check here, if we are adding it in `_secure_channel_with_interceptor` and `_insecure_channel_with_interceptor`.

I am still trying to think this through though.

Basically if we want to allow them to do:

```python
with grpc.insecure_channel('localhost:%d' %(GRPC_PORT)) as channel:
    Pin(service='my-grpc').onto(channel)
```

Reply:
Actually, it's kind of hard to achieve as the channel is recreated before being sent back to the client (c.f. [this](https://github.com/DataDog/dd-trace-py/pull/641/files#diff-6d1da44ed5c4cb935bacc4c9ccd06b3bR43)).

Each time the client adds a new interceptor, the channel is wrapped into another one. How can we get the real instance ?

Reply:
*/me groans*